File: benchmark/head_to_head.py
#!/usr/bin/env python3
"""
Head-to-head: StochTree-Ex (Elixir/Rust) vs StochTree Python 0.4.0
Runs matching BART configurations and reports wall time + RMSE.

Usage:
    ~/projects/learn_erl/python-env/bin/python benchmark/head_to_head.py
"""

import numpy as np
import time
import json
from stochtree import BARTModel


def friedman1(n, p_noise=5, seed=42):
    rng = np.random.RandomState(seed)
    p = 5 + p_noise
    X = rng.uniform(0, 1, (n, p))
    y = (10 * np.sin(np.pi * X[:, 0] * X[:, 1])
         + 20 * (X[:, 2] - 0.5) ** 2
         + 10 * X[:, 3]
         + 5 * X[:, 4]
         + rng.normal(0, 1, n))
    return X, y


def friedman2(n, seed=42):
    rng = np.random.RandomState(seed)
    X = np.column_stack([
        rng.uniform(0, 100, n),
        rng.uniform(40 * np.pi, 560 * np.pi, n),
        rng.uniform(0, 1, n),
        rng.uniform(1, 11, n),
    ])
    y = np.sqrt(X[:, 0] ** 2 + (X[:, 1] * X[:, 2] - 1 / (X[:, 1] * X[:, 3])) ** 2)
    y += rng.normal(0, y.std() * 0.1, n)
    return X, y


def simple_linear(n, seed=42):
    rng = np.random.RandomState(seed)
    X = rng.uniform(0, 10, (n, 3))
    y = 3 * X[:, 0] + rng.normal(0, 0.5, n)
    return X, y


def high_p(n, p, seed=42):
    """y = 2*x1 + noise, p-1 noise features."""
    rng = np.random.RandomState(seed)
    X = rng.uniform(0, 1, (n, p))
    y = 2.0 * X[:, 0] + rng.normal(0, 0.5, n)
    return X, y


def run_python_bart(name, X, y, num_trees=200, num_gfr=10, num_mcmc=100):
    model = BARTModel()
    t0 = time.time()
    model.sample(
        X_train=X, y_train=y,
        X_test=X,
        num_gfr=num_gfr,
        num_mcmc=num_mcmc,
        mean_forest_params={"num_trees": num_trees},
    )
    elapsed = time.time() - t0

    y_hat = model.y_hat_test
    if y_hat.shape[0] == len(y):
        y_hat = y_hat.T
    y_hat = y_hat[-num_mcmc:]
    y_pred = y_hat.mean(axis=0)
    rmse = np.sqrt(np.mean((y_pred - y) ** 2))

    return {"name": name, "time_ms": int(elapsed * 1000), "rmse": round(float(rmse), 4)}


def main():
    results = []
    print("=" * 70)
    print("StochTree Python 0.4.0 — Head-to-Head Reference")
    print("=" * 70)
    print()

    tests = [
        ("smoke (n=1K, p=10)", lambda: friedman1(1000, p_noise=5), 200, 100, 100),
        ("friedman2 (n=500, p=4)", lambda: friedman2(500), 200, 10, 100),
        ("simple-linear (n=200, p=3)", lambda: simple_linear(200), 50, 5, 50),
        ("high-p-100 (n=1K, p=100)", lambda: high_p(1000, 100), 200, 100, 100),
        ("extreme-p-500 (n=1K, p=500)", lambda: high_p(1000, 500), 200, 100, 100),
        ("medium-n (n=5K, p=10)", lambda: friedman1(5000, p_noise=5), 200, 100, 100),
        ("large-n (n=10K, p=10)", lambda: friedman1(10000, p_noise=5), 200, 100, 100),
    ]

    print(f"{'Test':<35} {'Time':>10} {'RMSE':>10}")
    print("-" * 60)

    for name, gen_fn, trees, gfr, mcmc in tests:
        X, y = gen_fn()
        r = run_python_bart(name, X, y, trees, gfr, mcmc)
        results.append(r)
        print(f"  {name:<33} {r['time_ms']:>8}ms  {r['rmse']:>8}")

    print()
    print("=" * 70)

    with open("benchmark/python_head_to_head.json", "w") as f:
        json.dump(results, f, indent=2)
    print("Saved to benchmark/python_head_to_head.json")


if __name__ == "__main__":
    main()

File: benchmark/head_to_head.exs
# Head-to-head: StochTree-Ex vs StochTree Python 0.4.0
# Runs matching BART configs and reports wall time + RMSE.
#
# Usage:
#   cd stochtree_ex && elixir --erl "+sbt tnnps" -S mix run benchmark/head_to_head.exs
#
# Compare with:
#   ~/projects/learn_erl/python-env/bin/python benchmark/head_to_head.py

defmodule HeadToHead do
  def friedman1(n, p_noise \\ 5, seed \\ 42) do
    :rand.seed(:exsss, {seed, seed + 1, seed + 2})
    p = 5 + p_noise
    x = for _ <- 1..n, do: for(_ <- 1..p, do: :rand.uniform())
    y = Enum.map(x, fn row ->
      [x1, x2, x3, x4, x5 | _] = row
      10 * :math.sin(:math.pi() * x1 * x2) +
        20 * :math.pow(x3 - 0.5, 2) +
        10 * x4 + 5 * x5 + :rand.normal()
    end)
    {x, y}
  end

  def friedman2(n, seed \\ 42) do
    :rand.seed(:exsss, {seed, seed + 1, seed + 2})
    x = for _ <- 1..n do
      [
        :rand.uniform() * 100,
        :rand.uniform() * (560 - 40) * :math.pi() + 40 * :math.pi(),
        :rand.uniform(),
        :rand.uniform() * 10 + 1
      ]
    end
    y = Enum.map(x, fn [x1, x2, x3, x4] ->
      val = :math.sqrt(x1 * x1 + :math.pow(x2 * x3 - 1 / (x2 * x4), 2))
      val + :rand.normal() * val * 0.1
    end)
    {x, y}
  end

  def simple_linear(n, seed \\ 42) do
    :rand.seed(:exsss, {seed, seed + 1, seed + 2})
    x = for _ <- 1..n, do: for(_ <- 1..3, do: :rand.uniform() * 10)
    y = Enum.map(x, fn [x1 | _] -> 3.0 * x1 + :rand.normal() * 0.5 end)
    {x, y}
  end

  def high_p(n, p, seed \\ 42) do
    :rand.seed(:exsss, {seed, seed + 1, seed + 2})
    x = for _ <- 1..n, do: for(_ <- 1..p, do: :rand.uniform())
    y = Enum.map(x, fn row -> Enum.at(row, 0) * 2.0 + :rand.normal() * 0.5 end)
    {x, y}
  end

  def rmse(y_true, y_pred) do
    n = length(y_true)
    ss = Enum.zip(y_true, y_pred)
         |> Enum.map(fn {a, b} -> (a - b) * (a - b) end)
         |> Enum.sum()
    :math.sqrt(ss / n)
  end

  def run_test(name, x, y, opts) do
    t0 = System.monotonic_time(:millisecond)
    {forest, _sigmas} = StochTree.BART.fit(x, y, opts)
    elapsed = System.monotonic_time(:millisecond) - t0

    %{mean: preds} = StochTree.predict(forest, x)
    r = rmse(y, preds)
    {name, elapsed, Float.round(r, 4)}
  end
end

IO.puts(String.duplicate("=", 70))
IO.puts("StochTree-Ex (Elixir/Rust) — Head-to-Head")
IO.puts("#{System.schedulers_online()} schedulers")
IO.puts(String.duplicate("=", 70))
IO.puts("")

tests = [
  {"smoke (n=1K, p=10)", fn -> HeadToHead.friedman1(1000, 5) end,
   [num_trees: 200, num_gfr: 100, num_mcmc: 100, seed: 42]},
  {"friedman2 (n=500, p=4)", fn -> HeadToHead.friedman2(500) end,
   [num_trees: 200, num_gfr: 10, num_mcmc: 100, seed: 42]},
  {"simple-linear (n=200, p=3)", fn -> HeadToHead.simple_linear(200) end,
   [num_trees: 50, num_gfr: 5, num_mcmc: 50, seed: 42]},
  {"high-p-100 (n=1K, p=100)", fn -> HeadToHead.high_p(1000, 100) end,
   [num_trees: 200, num_gfr: 100, num_mcmc: 100, seed: 42]},
  {"extreme-p-500 (n=1K, p=500)", fn -> HeadToHead.high_p(1000, 500) end,
   [num_trees: 200, num_gfr: 100, num_mcmc: 100, seed: 42]},
  {"medium-n (n=5K, p=10)", fn -> HeadToHead.friedman1(5000, 5) end,
   [num_trees: 200, num_gfr: 100, num_mcmc: 100, seed: 42]},
  {"large-n (n=10K, p=10)", fn -> HeadToHead.friedman1(10000, 5) end,
   [num_trees: 200, num_gfr: 100, num_mcmc: 100, seed: 42]},
]

header = String.pad_trailing("Test", 35) <> String.pad_leading("Time", 10) <> String.pad_leading("RMSE", 10)
IO.puts(header)
IO.puts(String.duplicate("-", 60))

results = Enum.map(tests, fn {name, gen_fn, opts} ->
  {x, y} = gen_fn.()
  {name, elapsed, rmse} = HeadToHead.run_test(name, x, y, opts)
  line = "  " <> String.pad_trailing(name, 33) <>
    String.pad_leading("#{elapsed}ms", 8) <>
    String.pad_leading("#{rmse}", 10)
  IO.puts(line)
  %{name: name, time_ms: elapsed, rmse: rmse}
end)

IO.puts("")
IO.puts(String.duplicate("=", 70))

json = Jason.encode!(results, pretty: true)
File.write!("benchmark/elixir_head_to_head.json", json)
IO.puts("Saved to benchmark/elixir_head_to_head.json")
